ArduCopter/GCS_Mavlink.cpp

```diff
@@ -1021,6 +1016,38 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_pause_continue(const mavlink_comma
     return MAV_RESULT_FAILED;
 }
 
+MAV_RESULT GCS_MAVLINK_Copter::handle_command_nav_takeoff(const mavlink_command_int_t &packet)
+{
+    // param3 : horizontal navigation by pilot acceptable
+    // param4 : yaw angle   (not supported)
+    // param5 : latitude    (not supported)
+    // param6 : longitude   (not supported)
+    // param7 : altitude [metres]
+
+    // create takeoff location from command int
+    Location takeoff_location;
+    if (!location_from_command_t(packet, takeoff_location)) {
+        return MAV_RESULT_DENIED;
+    }
+
+    // sanitize the takeoff location
+    takeoff_location.sanitize(copter.current_loc);
```

Comment:
I don't think we should add this sanitize because if the message provides and altitude of zero this would be changed to the vehicle's current altitude above home which could be quite different.

Reply:
@rmackay9 Thanks for your feedback.
But without that line, takeoff with terrain fails with the below command even at the starting location:
`command_int GLOBAL_TERRAIN_ALT MAV_CMD_NAV_TAKEOFF 0 0 0 0 0 0 0 0 10`

I noticed this when I take off from a different location to check if the terrain takeoff works.
There is a close location near the SITL starting location which is (-35.36266697 149.16138212) which has some terrain altitude from the starting location.
So I took off at the SITL start location.
Takeoff to 10 meters using `takeoff 10` (uses COMMAND_LONG in the background).
It works as it should so ✅ for the COMMAND_LONG part 🙂.
I went to the location (-35.36266697 149.16138212) while maintaining the 10 meters of altitude.
When I reached that location I landed.
Waited for the vehicle to disarm.
After arming (home location changed to the current location), I deliberately changed the EKF and home location to the initial SITL location by right-clicking the MAVProxy map and clicking both `set home with height`, `set origin with height`.
The current implementation, when I send the command `command_int GLOBAL_TERRAIN_ALT MAV_CMD_NAV_TAKEOFF 0 0 0 0 0 0 0 0 10`, it works as it should, reaches 10 meters terrain altitude.

So I removed the line as you suggested and did some experiments.
But with your suggestion, it fails to take-off, off at the initial SITL start location and also at the new location.
Any suggestions?